petsctools/monitor.py

```diff
@@ -0,0 +1,16 @@
+import abc
+
+from petsc4py import PETSc
+
+
+class AbstractKSPMonitorFunction(abc.ABC):
```

Comment:
I think this is a good idea, but there's nothing that makes python monitors in particular special. If we're doing ABCs we should do as much of the suite of possible python types as possible, e.g. KSP, PC, SNES, Monitor, Linesearch etc. (see for example the KSP and PC examples in the petsc4py docs)

Reply:
Some of the protocols are here: https://petsc.org/main/petsc4py/petsc_python_types.html
I think there's some discussion about what should be an abstract method (i.e. throw an error if a child class doesn't implement it), and what are just blank methods to show what you can customise.

Reply:
What I really want to have in petsctools are implementations of useful monitor functions and such. I'm less bothered about abstract classes and since petsc4py advertise most of the interfaces anyway it might be unnecessary here.